kobo-uncaged/kunc/kunc.go

```
@@ -0,0 +1,296 @@
+		}
+		bcd := uc.BookCountDetails{
+			UUID:         md.UUID,
+			Lpath:        md.Lpath,
+			LastModified: lastMod,
+		}
+		bcd.Extension = filepath.Ext(md.Lpath)
+		bc = append(bc, bcd)
+	}
+	//spew.Dump(bc)
+	return bc
+}
+
+// GetMetadataList sends complete metadata for the books listed in lpaths, or for
+// all books on device if lpaths is empty
+func (ku *koboUncaged) GetMetadataList(books []uc.BookID) []map[string]interface{} {
+	//spew.Dump(ku.k.MetadataMap)
+	//spew.Dump(books)
+	mdList := []map[string]interface{}{}
+	if len(books) > 0 {
+		for _, bk := range books {
+			cid := util.LpathToContentID(bk.Lpath, string(ku.k.ContentIDprefix))
+			fmt.Println(cid)
+			md := map[string]interface{}{}
+			//spew.Dump(ku.k.MetadataMap[cid])
+			mapstructure.Decode(ku.k.MetadataMap[cid], &md)
+			mdList = append(mdList, md)
+		}
+	} else {
+		for _, kmd := range ku.k.MetadataMap {
+			md := map[string]interface{}{}
+			//spew.Dump(kmd)
+			mapstructure.Decode(kmd, &md)
+			mdList = append(mdList, md)
+		}
+	}
+	return mdList
+}
+
+// GetDeviceInfo asks the client for information about the drive info to use
+func (ku *koboUncaged) GetDeviceInfo() uc.DeviceInfo {
+	return ku.k.DriveInfo
+}
+
+// SetDeviceInfo sets the new device info, as comes from calibre. Only the nested
+// struct DevInfo is modified.
+func (ku *koboUncaged) SetDeviceInfo(devInfo uc.DeviceInfo) {
+	ku.k.DriveInfo = devInfo
+	ku.k.SaveDeviceInfo()
+}
+
+// UpdateMetadata instructs the client to update their metadata according to the
+// new slice of metadata maps
+func (ku *koboUncaged) UpdateMetadata(mdList []map[string]interface{}) {
+	for _, md := range mdList {
+		koboMD := device.CreateKoboMetadata()
+		mapstructure.Decode(md, &koboMD)
+		koboMD.Thumbnail = nil
+		cid := util.LpathToContentID(koboMD.Lpath, string(ku.k.ContentIDprefix))
+		ku.k.MetadataMap[cid] = koboMD
+		ku.k.UpdatedMetadata = append(ku.k.UpdatedMetadata, cid)
+	}
+	ku.k.WriteMDfile()
+	ku.k.WriteUpdateMDfile()
```

Is there a reason for ignoring the errors here?

---

Yeah, because the current interface says we don't return an error.

I think the interface as a whole needs to be looked at again, but that is an UNCaGED matter, and outside the scope of this PR